include/dr/detail/communicator.hpp

```diff
@@ -56,26 +56,34 @@ class communicator {
                  mpi_comm_);
   }
 
-  void gather(const void *src, void *dst, std::size_t size,
+  void gather(const void *src, void *dst, std::size_t count,
               std::size_t root) const {
-    MPI_Gather(src, size, MPI_BYTE, dst, size, MPI_BYTE, root, mpi_comm_);
+    MPI_Gather(src, count, MPI_BYTE, dst, count, MPI_BYTE, root, mpi_comm_);
   }
 
   template <typename T>
   void gather(const T &src, std::span<T> dst, std::size_t root) const {
-    assert(dst.size() >= size());
-    gather(&src, dst.data(), sizeof(T), root);
+    assert(rng::size(dst) >= size_);
+    gather(&src, rng::data(dst), sizeof(T), root);
   }
 
-  void all_gather(const void *src, void *dst, int sz) const {
+  template <typename T>
+  void all_gather(const T *src, T *dst, std::size_t count) const {
     // Gather size elements from each rank
-    MPI_Allgather(src, sz, MPI_BYTE, dst, sz, MPI_BYTE, mpi_comm_);
+    MPI_Allgather(src, count * sizeof(T), MPI_BYTE, dst, count * sizeof(T),
+                  MPI_BYTE, mpi_comm_);
   }
 
   template <typename T>
   void all_gather(const T &src, std::vector<T> &dst) const {
-    assert(dst.size() >= size());
-    all_gather(&src, dst.data(), sizeof(T));
+    assert(rng::size(dst) >= size_);
+    all_gather(&src, rng::data(dst), 1);
+  }
+
+  template <typename T>
+  void all_gather(std::vector<T> &src, std::vector<T> &dst) const {
```

Comment:
one of vectors can be const

and vectors could be made `rng::contiguous_range` which should work for `vector` and any other class having data and size CPOs